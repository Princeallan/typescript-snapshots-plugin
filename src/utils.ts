import * as ts_module from "typescript/lib/tsserverlibrary";


/**
 * Find closest node at given position
 *
 * @export
 * @param ts
 * @param sourceFile
 * @param position
 * @returns
 */
export function findNodeAtPosition(ts: typeof ts_module, sourceFile: ts.SourceFile, position: number): ts_module.Node | undefined {
    function find(node: ts.Node): ts_module.Node | undefined {
        if (position >= node.getStart(sourceFile) && position <= node.getEnd()) {
            return ts.forEachChild(node, find) || node;
        }
        return;
    }
    return ts.forEachChild(sourceFile, find);
}

/**
 * Check if given node is the call expression with matching identifiers.
 * Main use is the checking is the node is toMatchSnapshot identifier call
 *
 * @export
 * @param ts
 * @param node
 * @param identifiers
 * @returns
 */
export function isMatchingCallExpression(ts: typeof ts_module, node: ts_module.Node, identifiers: string[]): node is ts_module.CallExpression {
    if (!ts.isCallExpression(node)) {
        return false;
    }
    const text = node.expression.getText();
    return identifiers.indexOf(text) !== -1;
}

/**
 * Check if given node is identifiers and match any of given identifiers
 *
 * @export
 * @param ts
 * @param node
 * @param identifiers
 * @returns
 */
export function isMatchingIdentifier(ts: typeof ts_module, node: ts_module.Node, identifiers: string[]): node is ts_module.Identifier {
    if (!ts.isIdentifier(node)) {
        return false;
    }
    return identifiers.indexOf(node.getText()) !== -1;
}

export interface TestBlockInfo {
    /**
     * Parent + current test block names
     */
    blockNames: string[]
    /**
     * Last test block node
     */
    lastNode: ts.Node;
}

/**
 * Return all parent test blocks
 *
 * @export
 * @param ts
 * @param sourceFile
 * @param testBlockIdentifiers
 * @param lookupUntil
 * @returns
 */
export function getParentTestBlocks(ts: typeof ts_module, sourceFile: ts.SourceFile, testBlockIdentifiers: string[], lookupUntil: number): TestBlockInfo | undefined {
    const blocks: string[] = [];
    let lastBlockNode: ts.Node | undefined;
    function find(node: ts.Node): void {
        if (lookupUntil >= node.getStart() && lookupUntil <= node.getEnd()) {
            if (isMatchingCallExpression(ts, node, testBlockIdentifiers) && node.arguments.length >= 1) {
                if (ts.isStringLiteral(node.arguments[0]) || ts.isNoSubstitutionTemplateLiteral(node.arguments[0])) {
                    lastBlockNode = node;
                    blocks.push((node.arguments[0] as ts_module.StringLiteral).text);
                    ts.forEachChild(node, find);
                }
            } else {
                ts.forEachChild(node, find);
            }
        }
    }
    find(sourceFile);
    if (!lastBlockNode) {
        return undefined;
    }
    return {
        blockNames: blocks,
        lastNode: lastBlockNode
    };
}

/**
 * Calculate count of given identifiers in the bloc
 *
 * @export
 * @param ts
 * @param block
 * @param lookupUntil
 * @returns
 */
export function getCountOfIdentifiersInBlock(
    ts: typeof ts_module,
    block: ts_module.Node,
    identifiers: string[],
    lookupUntil: number
): { anonymousCalls: number; namedCalls: { [key: string]: number } } {
    let anonymousCalls = 0;
    const namedCalls: { [key: string]: number } = { };
    function visit(node: ts.Node): void {
        if (ts.isIdentifier(node) && identifiers.indexOf(node.text) !== -1 && lookupUntil >= node.getStart()) {
            if (node.parent && node.parent.parent && ts.isCallExpression(node.parent.parent)) {
                const customName = node.parent.parent.arguments[0];
                if (customName && ts.isStringLiteralLike(customName)) {
                    const snapshotName = customName.text;
                    namedCalls[snapshotName] = namedCalls[snapshotName] ? namedCalls[snapshotName] + 1 : 1;
                } else {
                    anonymousCalls++;
                }
            }
        }
        ts.forEachChild(node, visit);
    }
    ts.forEachChild(block, visit);
    return {
        anonymousCalls,
        namedCalls,
    };
}

export interface SnapshotDefinition {
    /**
     * Snapshot export name
     */
    name: string;
    /**
     * Snapshot source
     */
    snapshot: string;
    /**
     * Snapshot position
     */
    position: number;
    /**
     * Length of snapshot export
     */
    length: number;
}

/**
 * Parse snapshot-like file and return snapshot definitions
 *
 * @export
 * @param ts
 * @param source
 * @returns
 */
export function parseSnapshotFile(ts: typeof ts_module, source: string): SnapshotDefinition[] {
    try {
        const file = ts.createSourceFile("temp.ts", source, ts.ScriptTarget.Latest, true, ts.ScriptKind.TSX);
        const blocks: SnapshotDefinition[] = [];

        const visit = (node: ts.Node) => {
            if (ts.isBinaryExpression(node) &&
                ts.isElementAccessExpression(node.left) &&
                node.left.argumentExpression &&
                (ts.isStringLiteral(node.left.argumentExpression) || ts.isNoSubstitutionTemplateLiteral(node.left.argumentExpression)) &&
                (ts.isStringLiteral(node.right) || ts.isNoSubstitutionTemplateLiteral(node.right))
            ) {
                blocks.push({
                    name: node.left.argumentExpression.text,
                    snapshot: node.right.text,
                    position: node.getStart(),
                    length: node.getText().length
                });
            } else {
                ts.forEachChild(node, visit);
            }
        }
        ts.forEachChild(file, visit);
        return blocks;
    } catch {
        return [];
    }
}