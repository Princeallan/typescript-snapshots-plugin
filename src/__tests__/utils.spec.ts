import * as ts from "typescript";
import { findNodeAtPosition, isMatchingCallExpression, isMatchingIdentifier, getParentTestBlocks, getCountOfIdentifiersInBlock, parseSnapshotFile } from "../utils";

const source = `
it("test", () => {
    expect(a).toMatchSnapshot();
});

it.only("test2", () => {
    expect(a).toMatchSnapshot();
});

describe("a", () => {
    it("test2", () => {
        expect(b).toMatchSnapshot();
        // invalid
        expect(c).toMatchSnapshot(
    });
});

describe();
describe("test", () => {});
it("test", () => {
    expect(a).toBe(1);
});

describe("valid", () => {
    it("test1", () => {
        expect(a).toMatchSnapshot();
        expect(a).toMatchSnapshot();
        expect(a).toMatchSnapshot();
    });

    describe("inner", () => {
        it("test2", () => {
            expect(b).toMatchSnapshot();
        });

        it("test3", () => {
            expect(b).toMatchSnapshot();
        });
    });
});
`;
const file = ts.createSourceFile("a.ts", source, ts.ScriptTarget.Latest, true, ts.ScriptKind.TS);

describe("findNodeAtPosition", () => {
    it("Returns node at position", () => {
        let node = findNodeAtPosition(ts as any, file, ts.getPositionOfLineAndCharacter(file, 1, 2));
        expect(node!.getText()).toBe("it");

        node = findNodeAtPosition(ts as any, file, ts.getPositionOfLineAndCharacter(file, 2, 7));
        expect(node!.getText()).toBe("expect");

        node = findNodeAtPosition(ts as any, file, ts.getPositionOfLineAndCharacter(file, 2, 25));
        expect(node!.getText()).toBe("toMatchSnapshot");

        node = findNodeAtPosition(ts as any, file, ts.getPositionOfLineAndCharacter(file, 4, 0));
        expect(node).toBeUndefined();

        node = findNodeAtPosition(ts as any, file, ts.getPositionOfLineAndCharacter(file, 11, 27));
        expect(node!.getText()).toBe("toMatchSnapshot");
    })
});

describe("isMatchingIdentifier", () => {
    it("Returns true if node is identifier and matches given identifiers", () => {
        let node = findNodeAtPosition(ts as any, file, ts.getPositionOfLineAndCharacter(file, 1, 2));
        expect(isMatchingIdentifier(ts as any, node!, ["it"])).toBeTruthy();
        expect(isMatchingIdentifier(ts as any, node!, ["its"])).toBeFalsy();

        node = findNodeAtPosition(ts as any, file, ts.getPositionOfLineAndCharacter(file, 2, 21));
        expect(isMatchingIdentifier(ts as any, node!, ["toMatchSnapshot"])).toBeTruthy();
        expect(isMatchingIdentifier(ts as any, node!, ["toMatchSnapsht"])).toBeFalsy();
    });
});

describe("isMatchingCallExpression", () => {
    it("Returns true if node is call expression and matches given identifiers", () => {
        let node = findNodeAtPosition(ts as any, file, ts.getPositionOfLineAndCharacter(file, 1, 2));
        expect(isMatchingCallExpression(ts as any, node!.parent!, ["it"])).toBeTruthy();
        expect(isMatchingCallExpression(ts as any, node!.parent!, ["its"])).toBeFalsy();

        node = findNodeAtPosition(ts as any, file, ts.getPositionOfLineAndCharacter(file, 5, 5));
        // identifier -> property access -> call exp
        expect(isMatchingCallExpression(ts as any, node!.parent!.parent!, ["it.only"])).toBeTruthy();
        expect(isMatchingCallExpression(ts as any, node!.parent!.parent!, ["it"])).toBeFalsy();

        node = findNodeAtPosition(ts as any, file, ts.getPositionOfLineAndCharacter(file, 9, 4));
        expect(isMatchingCallExpression(ts as any, node!.parent!, ["describe"])).toBeTruthy();
    });
})

describe("getParentTestBlocks", () => {
    it("Returns last test block and all parent names", () => {
        let res = getParentTestBlocks(ts as any, file, ["describe", "it"], ts.getPositionOfLineAndCharacter(file, 2, 17));
        expect(res).toBeDefined();
        expect(res!.blockNames).toEqual(["test"]);
        expect(res!.lastNode).toBeDefined();
        expect(res!.lastNode.getText()).toMatchSnapshot();

        res = getParentTestBlocks(ts as any, file, ["describe", "it"], ts.getPositionOfLineAndCharacter(file, 6, 20));
        expect(res).toBeUndefined();
        res = getParentTestBlocks(ts as any, file, ["describe", "it", "it.only"], ts.getPositionOfLineAndCharacter(file, 6, 20));
        expect(res).toBeDefined();
        expect(res!.blockNames).toEqual(["test2"]);
        expect(res!.lastNode.getText()).toMatchSnapshot();

        res = getParentTestBlocks(ts as any, file, ["describe", "it"], ts.getPositionOfLineAndCharacter(file, 11, 22));
        expect(res).toBeDefined();
        expect(res!.blockNames).toEqual(["a", "test2"])
        expect(res!.lastNode.getText()).toMatchSnapshot();

        res = getParentTestBlocks(ts as any, file, ["describe", "it"], ts.getPositionOfLineAndCharacter(file, 32, 32));
        expect(res).toBeDefined();
        expect(res!.blockNames).toEqual(["valid", "inner", "test2"]);
        expect(res!.lastNode.getText()).toMatchSnapshot();

        res = getParentTestBlocks(ts as any, file, ["describe", "it"], ts.getPositionOfLineAndCharacter(file, 36, 36));
        expect(res).toBeDefined();
        expect(res!.blockNames).toEqual(["valid", "inner", "test3"]);
        expect(res!.lastNode.getText()).toMatchSnapshot();

        res = getParentTestBlocks(ts as any, file, ["describe", "it"], ts.getPositionOfLineAndCharacter(file, 27, 27));
        expect(res).toBeDefined();
        expect(res!.blockNames).toEqual(["valid", "test1"]);
        expect(res!.lastNode.getText()).toMatchSnapshot();
    });
});

describe("getCountOfIdentifiersInBlock", () => {
    it("Returns number of identifiers in given block", () => {
        let node = findNodeAtPosition(ts as any, file, ts.getPositionOfLineAndCharacter(file, 1, 1));
        expect(getCountOfIdentifiersInBlock(ts as any, node!.parent!, ["toMatchSnapshot"], ts.getPositionOfLineAndCharacter(file, 2, 29))).toBe(1);
        expect(getCountOfIdentifiersInBlock(ts as any, node!.parent!, ["toMatchSnapshot"], ts.getPositionOfLineAndCharacter(file, 2, 13))).toBe(0);

        node = findNodeAtPosition(ts as any, file, ts.getPositionOfLineAndCharacter(file, 10, 5));
        expect(getCountOfIdentifiersInBlock(ts as any, node!.parent!, ["toMatchSnapshot"], ts.getPositionOfLineAndCharacter(file, 12, 9))).toBe(1);
        expect(getCountOfIdentifiersInBlock(ts as any, node!.parent!, ["toMatchSnapshot"], ts.getPositionOfLineAndCharacter(file, 13, 17))).toBe(1);
        expect(getCountOfIdentifiersInBlock(ts as any, node!.parent!, ["toMatchSnapshot"], ts.getPositionOfLineAndCharacter(file, 13, 34))).toBe(2);

        node = findNodeAtPosition(ts as any, file, ts.getPositionOfLineAndCharacter(file, 24, 5));
        expect(getCountOfIdentifiersInBlock(ts as any, node!.parent!, ["toMatchSnapshot"], ts.getPositionOfLineAndCharacter(file, 24, 23))).toBe(0);
        expect(getCountOfIdentifiersInBlock(ts as any, node!.parent!, ["toMatchSnapshot"], ts.getPositionOfLineAndCharacter(file, 25, 9))).toBe(0);
        expect(getCountOfIdentifiersInBlock(ts as any, node!.parent!, ["toMatchSnapshot"], ts.getPositionOfLineAndCharacter(file, 26, 9))).toBe(1);
        expect(getCountOfIdentifiersInBlock(ts as any, node!.parent!, ["toMatchSnapshot"], ts.getPositionOfLineAndCharacter(file, 27, 9))).toBe(2);
        expect(getCountOfIdentifiersInBlock(ts as any, node!.parent!, ["toMatchSnapshot"], ts.getPositionOfLineAndCharacter(file, 27, 18))).toBe(2);
        expect(getCountOfIdentifiersInBlock(ts as any, node!.parent!, ["toMatchSnapshot"], ts.getPositionOfLineAndCharacter(file, 28, 5))).toBe(3);
    });
});

describe("parseSnapshotFile", () => {
    const snapshot = `
        exports["abc"] = "abc";

        exports[\`abc\`] = \`
        abc,
        def
        \`;

        module.exports["bb"] = "abc";

        exports["def"] = "";
    `;

    const empty = ``;

    it("Returns empty array for empty snapshot", () => {
        expect(parseSnapshotFile(ts as any, empty).length).toBe(0);
    });

    it("Returns snapshot definitions", () => {
        const res = parseSnapshotFile(ts as any, snapshot);
        expect(res).toMatchSnapshot();
    });
})