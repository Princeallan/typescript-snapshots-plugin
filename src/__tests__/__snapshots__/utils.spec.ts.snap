// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getParentTestBlocks Returns last test block and all parent names 1`] = `
"it(\\"test\\", () => {
    expect(a).toMatchSnapshot();
})"
`;

exports[`getParentTestBlocks Returns last test block and all parent names 2`] = `
"it.only(\\"test2\\", () => {
    expect(a).toMatchSnapshot();
})"
`;

exports[`getParentTestBlocks Returns last test block and all parent names 3`] = `
"it(\\"test2\\", () => {
        expect(b).toMatchSnapshot();
        // invalid
        expect(c).toMatchSnapshot(
    })"
`;

exports[`getParentTestBlocks Returns last test block and all parent names 4`] = `
"it(\\"test2\\", () => {
            expect(b).toMatchSnapshot();
        })"
`;

exports[`getParentTestBlocks Returns last test block and all parent names 5`] = `
"it(\\"test3\\", () => {
            expect(b).toMatchSnapshot();
        })"
`;

exports[`getParentTestBlocks Returns last test block and all parent names 6`] = `
"it(\\"test1\\", () => {
        expect(a).toMatchSnapshot();
        expect(a).toMatchSnapshot();
        expect(a).toMatchSnapshot();
    })"
`;

exports[`parseSnapshotFile Returns snapshot definitions 1`] = `
Array [
  Object {
    "length": 22,
    "name": "abc",
    "position": 9,
    "snapshot": "abc",
  },
  Object {
    "length": 53,
    "name": "abc",
    "position": 42,
    "snapshot": "
        abc,
        def
        ",
  },
  Object {
    "length": 28,
    "name": "bb",
    "position": 106,
    "snapshot": "abc",
  },
  Object {
    "length": 19,
    "name": "def",
    "position": 145,
    "snapshot": "",
  },
]
`;
